### PR TITLE
improve object var symantics

### PR DESCRIPTION
**Pull request by adhami3310**
@benedikt-bartscher made some changes here, let me know if any more is needed

### CODE REVIEW

**Comment by benedikt-bartscher** [comment]
Hi @adhami3310 thanks!
If i read the code correctly, attribute access is still not allowed for bare classes like this one:

```python
class Bare:
  @property
  def prop(self) -> int:
    return 4

@serializer(to=dict)
def serialize_bare(bare: Bare) -> dict[str, int]:
  return {"prop": bare.prop}
```

I would love to see a fallback to ObjectVar (a bit like https://github.com/reflex-dev/reflex/pull/3911).
Alternatively we could add a typing stub class like this:

```python
class IsObjectVar: ...

class Bare(IsObjectVar): ...
```

**Comment by adhami3310** [comment]
> attribute access is still not allowed for bare classes like this one

I believe so, I can add an additional check for has_serializer.

**Comment by benedikt-bartscher** [comment]
> > attribute access is still not allowed for bare classes like this one
> 
> I believe so, I can add an additional check for has_serializer.

I would suggest limiting this case to serializers which serialize to dict only.
Otherwise, it will be true for this:


```python
class MyCls: ...

@serializer(to=str)
def serialize_my_cls(my_cls: MyCls) -> str:
  return "hi"
```

**Comment by benedikt-bartscher** [comment]
Refs:
https://github.com/reflex-dev/reflex/pull/3911
https://github.com/reflex-dev/reflex/issues/4269
https://github.com/reflex-dev/reflex/pull/4268 (maybe we should pick some of the tests from that PR)

**Comment by adhami3310** [comment]
> I would suggest limiting this case to serializers which serialize to dict only.

Should be done!

> maybe we should pick some of the tests from that PR

we should, I will try to do that

**Review by benedikt-bartscher — COMMENTED** [review]
Thanks, @adhami3310 this looks great! I will test it against our code now

Edit: works great!

File: tests/units/vars/test_object.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    var = original_var.to(ObjectVar, type_)
+    assert var._var_type is type_
+
+    var = original_var.to(ObjectVar)
+    assert var._var_type is type_
+
+
+def test_typing() -> None:
+    if sys.version_info >= (3, 11):
```
[the remark this answers is not in the document]
Comment (adhami3310) [reply tests/units/vars/test_object.py:97]: I don't believe `assert_type` is available in `typing_extensions`, at least my editor didn't suggest that as an option :/

[the remark this answers is not in the document]
Comment (benedikt-bartscher) [reply tests/units/vars/test_object.py:97]: That's probably caused by the python version your lsp uses to interpret the code. It should be available https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.assert_type

[the remark this answers is not in the document]
Comment (adhami3310) [reply tests/units/vars/test_object.py]: I see, my editor must be doing something funny then

**Review by benedikt-bartscher — APPROVED** [review]

**Pull request merged**